File: basicAlgorithms/quick_sort_2_index.cpp
#include <iostream>
using namespace std;

void swap(int arr[], int i, int j)
{
    int temp;
    temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}
void quick_sort(int arr[], int left, int right)
{
    int right_index = right;
    int left_index = left + 1;
    int pivot = left;
    if (left >= right)
        return;

    while (right_index >= left_index)
    {
        if (arr[right_index] < arr[pivot] && arr[left_index] > arr[pivot])
        {
            swap(arr, left_index, right_index);
        }
        if (arr[left_index] <= arr[pivot])
            left_index++;
        if (arr[right_index] >= arr[pivot])
            right_index--;
    }
    swap(arr, right_index, pivot);
    int check_for_smaller_sub_array = right_index - left > right - right_index;

    if (check_for_smaller_sub_array)
    {
        quick_sort(arr, left, right_index - 1);
        quick_sort(arr, right_index + 1, right);
    }
    else
    {
        quick_sort(arr, right_index + 1, right);
        quick_sort(arr, left, right_index - 1);
    }
}
int main()
{
    int arr[] = {8, 54, 43, 2, 5, 3, 2};
    quick_sort(arr, 0, 6);
    for (int i = 0; i <= 6; i++)
        cout << arr[i] << "\t";
}

File: basicAlgorithms/quick_sort_2_index.py
def swap(array, i, j):
    array[i], array[j] = array[j], array[i]


def quickSort(array, left, right):
    if left >= right:
        return
    pivot = left
    leftIndex = left + 1
    rightIndex = right
    while rightIndex >= leftIndex:
        if array[leftIndex] > array[pivot] and array[rightIndex] < array[pivot]:
            swap(array, leftIndex, rightIndex)
        if array[leftIndex] <= array[pivot]:
            leftIndex += 1
        if array[rightIndex] >= array[pivot]:
            rightIndex -= 1
    swap(array, rightIndex, pivot)

    smallerSubArray = rightIndex - left >= right - rightIndex
    if smallerSubArray:
        quickSort(array, left, rightIndex - 1)
        quickSort(array, rightIndex+1, right)
    else:
        quickSort(array, rightIndex+1, right)
        quickSort(array, left, rightIndex - 1)


array = [8, 5, 2, 9, 5, 6, 3]
quickSort(array, 0, len(array) - 1)
for i in array:
    print(i, end="\t")
